### ISSUE TITLE
Will Web never be a "First-class citizen"? (and what is happening to the 3.x branch?)

### ISSUE THREAD

**Issue by PeterMarques** [godotengine/godot#88230]
### Tested versions

3.5.3

### System information

web

### Issue description

![image](https://github.com/godotengine/godot/assets/32851924/2b0d2e88-bc6e-4df6-a32b-68314baf169d)

I spoted a very low performance, on my godot based website, like 1fps.

Then, i remembered the test link, and i could reproduce that on the test link as well.

Its several years, Several.

And yet, godot on the web is Pure .....

Well, and there is the case for to abandon old computers, or use the 3.x branch, a branch that does not sees a release in a very long time, compared to 4.

I just think that its time for godot to either have good web implementation, a unifed one, on the 4 branch, that can attend ALL machines, and a implementation that can be performant on a Freaking blank page(All the time)???

Other Engines dithed the Web/WASM if i reccal. BUt at the sime time is not the Wasm problem per se..

I know this is a open source effort and i should help if i need help, not just point fingers.

But, cant be the only one to bother. 

In my sincere dreams, Godot is the new Flash Player. From Games to Single-page Sites, and all beyond. That should be the position it should hold.

But, it never delivered... Never.

Why? I just need to understand.

Thanks and sorry.

### Steps to reproduce

.

### Minimal reproduction project (MRP)

.

**Comment by clayjohn** [comment]
We definitely want the experience on web to reach parity with other platforms.

If you follow development you will see that we are making progress towards that too. Just this week we released 4.3 dev 3 which, among other things, includes support for single threaded web builds which really improves the experience for web games in a lot of situations https://godotengine.org/article/dev-snapshot-godot-4-3-dev-3/

Now, web is a complex beast, we are balancing many factors, including wanting to support modern features on all platforms and wrestling with poor WebGL support on certain browsers. There are many open issues in this repository dealing with specific issues that we are working on. 

To that end, what is most valuable for us is for users to open _specific_ bug reports reporting _specific_,  actionable bugs. 

Accordingly, I think it would be best if we closed this issue in favour of the others. If you want to have a general discussion about web support, please reach out on Discord, or the Godot forum at forum.godotengine.org

**Comment by PeterMarques** [comment]
Indeed. If i cant find the specific cause of the slow performance of the case mencioned, i will close it.

Like, right now its back to speed again. Fully performant.
![image](https://github.com/godotengine/godot/assets/32851924/9e9b04ed-bfd1-4c62-b1d0-ed47b65b4204)

And its the same brower, same user session, same computer.

And it is not the first time such a behaviou happend. Its not there was another process hangling, it was the godot tab itself using the machine.

And i tried several times, that moment

And now, its performant???

That inconsitancy, intermitency, sometimes its optmized, sometimes not? My machine did not changed, ergo WTF is happening? Some bug on the JIT? Idk, but a page cant sometimes work and sometimes not.

**Comment by Armynator** [comment]
This isn't a Godot problem, but a problem of your OS/Browser. On your first screenshot, a fallback CPU renderer was used by your web browser/OS, most likely because you have updated your graphics drivers during that session. The easiest way to fix that is to restart your computer.
As far as I know, this is a very common problem with all browsers on several Linux distributions. For some reason they can't render WebGL on the GPU without a full system restart after driver updates, so it either gets rendered on the CPU (Chromium does this iirc) or refuses to render right away (Firefox).

**Comment by Calinou** [comment]
We might be able to detect when a software fallback is used on startup and print a warning accordingly, based on `RenderingServer.get_video_adapter_name()` and `RenderingServer.get_video_adapter_vendor()` (`RenderingServer.get_video_adapter_type()` isn't available in OpenGL/WebGL).

**Comment by PeterMarques** [comment]
> We might be able to detect when a software fallback is used on startup and `alert()` a warning accordingly, based on `RenderingServer.get_video_adapter_name()` and `RenderingServer.get_video_adapter_vendor()` (`RenderingServer.get_video_adapter_type()` isn't available in OpenGL/WebGL).
> 
> There should be a CLI argument or environment variable to opt out of this, as software fallbacks are sometimes desired for automated testing.
> 
> This would also be worth doing for desktop platforms: [godotengine/godot-proposals#8071](https://github.com/godotengine/godot-proposals/issues/8071)

Thats a cool try.

But i dont think its really whats caused the issue, but i will have it ready on a testpage, maybe it really is. If i encounter that specific slowdown, i will try to compare the outputs of "normal" and "slow" performant loadouts, and report here.

But, i do not know how to force it to happen. Just wait. 

If it really is that, maybe there is a way to force-reload the page via JS and alocate corretly the driver.  But first, to determine the real cause.

Will test update here with the strings that i get on normal performance, and wait for a rundown to compare.

Thanks.

**Comment by Calinou** [comment]
> If it really is that, maybe there is a way to force-reload the page via JS and alocate corretly the driver. But first, to determine the real cause.

Switching graphics drivers requires restarting the browser entirely, which isn't something you can do from JavaScript.

**Comment by PeterMarques** [comment]
https://github.com/godotengine/godot/issues/70652#issuecomment-2224238731

**Comment by PeterMarques** [comment]
I am closing this issue since godot 4 on the web seen to have become more stable.
